### PR TITLE
Columbus-4 TerraSwap FCD Tx collector

### PR DESCRIPTION
**Pull request by jbamlee**
## Summary

- Collect Tx log data from the FCD

## Checklist

- [ ] Backward compatible?
- [ ] Test enough in your local environment?
- [ ] Add related test cases?

### CODE REVIEW

**Review by psy2848048 — APPROVED** [review]

File: collector/datastore/datastore.go [hunk 1]
```diff
@@ -405,7 +405,9 @@ func (store *dataStoreImpl) getTxResultFromTxHash(txHash string) (*txtypes.GetTx
 	resp, err := client.GetTx(context.Background(), req)
 	if err != nil {
 		// failover with lcd
-		resp, err = store.lcdClient.GetTx(txHash)
+		if store.lcdClient != nil {
```
Comment (jhlee-young) [inline collector/datastore/datastore.go:408]: can lcdClient be `nil` in any case?

File: collector/terra/fcd/permanent_store.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+func NewPermanentStore(db *gorm.DB) permanentStore {
+	return &permanentStoreImpl{db}
+}
+
+// FirstTxOf implements Columbus4Repository.
+func (p *permanentStoreImpl) FirstTxOf(addr string) (schemas.FcdTxLog, error) {
+	tx := schemas.FcdTxLog{}
+	if err := p.db.Where("address = ?", addr).Order("fcd_offset").Limit(1).Find(&tx).Error; err != nil {
+		return schemas.FcdTxLog{}, errors.Wrap(err, "col4RepoImpl.LatestTxOf")
```
Comment (jhlee-young) [inline collector/terra/fcd/permanent_store.go:23]: could you update wrapping message in this file?

File: collector/terra/fcd/collector.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func New(repo fcdRepo, store permanentStore) ColumbusCollector {
+	return &columbusFcdCollector{repo, store}
+}
+
+// hasCollected checks the absence of the previous transaction by examining the transactions collected from the FCD server
+//
+// NOTE: fcd returns transactions in reverse order of height
+func (c *columbusFcdCollector) hasCollected(addr string, minHeight uint32) (bool, error) {
```
Comment (jhlee-young) [inline collector/terra/fcd/collector.go:42]: whatabout deleting `minHeight`? seems not used

File: collector/terra/fcd/fcdepo.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	limit := fcd.FCD_TXS_MAX_LIMIT
+	offset := int(startOffset)
+	txs := []schemas.FcdTxLog{}
+	for ; ; offset = int(txs[len(txs)-1].FcdOffset) {
+		txsRes, err := rp.client.TxsOf(addr, fcd.FcdTxsReqQuery{Limit: &limit, Offset: &offset})
+		if err != nil {
+			if strings.Contains(err.Error(), fcd.STATUS_SERVER_ERROR.Error()) && 0 < errLimit {
+				errLimit--
+				time.Sleep(10 * time.Second)
+				fmt.Print(err.Error())
```
Comment (jhlee-young) [inline collector/terra/fcd/fcdepo.go:60]: please delete this line

File: collector/terra/fcd/permanent_store.go (hunk 2 above)
Comment (jbamlee) [reply collector/terra/fcd/permanent_store.go]: https://github.com/dezswap/cosmwasm-etl/pull/15/commits/c07da70eaeb89d08fa1de79b63e592ff1769b041

File: collector/terra/fcd/fcdepo.go (hunk 4 above)
Comment (jbamlee) [reply collector/terra/fcd/fcdepo.go]: https://github.com/dezswap/cosmwasm-etl/pull/15/commits/c07da70eaeb89d08fa1de79b63e592ff1769b041

File: collector/terra/fcd/collector.go (hunk 3 above)
Comment (jbamlee) [reply collector/terra/fcd/collector.go]: https://github.com/dezswap/cosmwasm-etl/pull/15/commits/c07da70eaeb89d08fa1de79b63e592ff1769b041

**Review by jhlee-young — APPROVED** [review]

**Pull request merged**